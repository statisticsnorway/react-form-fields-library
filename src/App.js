import React, { Component } from 'react'
import { DCFormField } from './main'
import { Checkbox, Divider, Form, Grid, Header } from 'semantic-ui-react'
import * as moment from 'moment'
import 'moment/locale/nb'

moment.locale('nb')

export const formComponents = {
  DCText: {
    component: 'DCText',
    name: 'myTextInput',
    displayName: 'DCText',
    description: 'A description for this input',
    error: '',
    warning: '',
    required: true,
    value: ''
  },
  DCBoolean: {
    component: 'DCBoolean',
    name: 'myBooleanInput',
    displayName: 'DCBoolean',
    description: 'A description for this input',
    value: false
  },
  DCNumber: {
    component: 'DCNumber',
    name: 'myNumberInput',
    displayName: 'DCNumber',
    description: 'A description for this input',
    error: '',
    warning: '',
    required: true,
    value: ''
  },
  DCRadio: {
    component: 'DCRadio',
    name: 'myRadioInput',
    displayName: 'DCRadio',
    description: 'A description for this input',
    error: '',
    warning: '',
    required: true,
    value: '',
    options: [
      {key: 'thisOption', text: 'This option', value: 'thisOption'},
      {key: 'thatOption', text: 'That option', value: 'thatOption'},
      {key: 'somethingElse', text: 'Something else', value: 'somethingElse'}
    ]
  },
  DCDate: {
    component: 'DCDate',
    name: 'myDateInput',
    displayName: 'DCDate',
    description: 'A description for this input',
    error: '',
    warning: '',
    required: true,
    value: ''
  },
  DCDropdownSingleSelect: {
    component: 'DCDropdown',
    name: 'myDropdownSingleSelectInput',
    displayName: 'DCDropdown',
    description: 'A description for this input',
    error: '',
    warning: '',
    required: true,
    value: '',
    endpoints: [
      'https://metadata.ssbmod.net/data/Role/',
      'https://metadata.ssbmod.net/data/Agent/'
    ],
    multiSelect: false
  },
  DCDropdownMultipleSelect: {
    component: 'DCDropdown',
    name: 'myDropdownMultipleSelectInput',
    displayName: 'DCDropdown (multiSelect)',
    description: 'A description for this input',
    error: '',
    warning: '',
    required: true,
    value: [],
    endpoints: [
      'https://metadata.ssbmod.net/data/Agent/',
      'https://metadata.ssbmod.net/data/Role/'
    ],
    multiSelect: true
  }
}

class App extends Component {
  constructor (props) {
    super(props)
    this.state = {
      ready: true,
      error: false,
      warning: false,
      urlError: false,
      networkError: false,
      formComponents: formComponents
    }
  }

  componentDidMount () {
    sessionStorage.clear()
  }

  handleCheckbox (name) {
    this.setState({
      [name]: !this.state[name]
    }, () => {
      if (this.state[name]) {
        this.setState({ready: false}, () => {
          const formComponents = JSON.parse(JSON.stringify(this.state.formComponents))

          Object.keys(formComponents).forEach(key => {
            formComponents[key][name] = name
          })

          this.setState({formComponents: formComponents}, () => {
            this.setState({ready: true})
          })
        })
      } else {
        this.setState({ready: false}, () => {
          const formComponents = JSON.parse(JSON.stringify(this.state.formComponents))

          Object.keys(formComponents).forEach(key => {
            formComponents[key][name] = ''
          })

          this.setState({formComponents: formComponents}, () => {
            this.setState({ready: true})
          })
        })
      }
    })
  }

  handleDropdownErrors (name) {
    this.setState({
      [name]: !this.state[name]
    }, () => {
      if (this.state[name]) {
        this.setState({ready: false}, () => {
          const formComponents = JSON.parse(JSON.stringify(this.state.formComponents))
          const errorEndpoints = []

          if (name === 'urlError') {
            errorEndpoints.push('https://metadata.ssbmod.net/data/Agentfda/')
            errorEndpoints.push('https://metadata.ssbmod.net/data/Roleyj/')
          }

          if (name === 'networkError') {
            errorEndpoints.push('https://metadatas.ssbmod.net/data/Agent/')
            errorEndpoints.push('https://metadataas.ssbmod.net/data/Role/')
          }

          Object.keys(formComponents).forEach(key => {
            if (key === 'DCDropdownSingleSelect' || key === 'DCDropdownMultipleSelect') {
              formComponents[key].endpoints = errorEndpoints
            }
          })

          this.setState({formComponents: formComponents}, () => {
            this.setState({ready: true})
          })
        })
      } else {
        this.setState({ready: false}, () => {
          const formComponents = JSON.parse(JSON.stringify(this.state.formComponents))
          const goodEndpoints = ['https://metadata.ssbmod.net/data/Agent/', 'https://metadata.ssbmod.net/data/Role/']

          Object.keys(formComponents).forEach(key => {
            if (key === 'DCDropdownSingleSelect' || key === 'DCDropdownMultipleSelect') {
              formComponents[key].endpoints = goodEndpoints
            }
          })

          this.setState({formComponents: formComponents}, () => {
            this.setState({ready: true})
          })
        })
      }
    })
  }

  render () {
    const {ready, warning, error, urlError, networkError, formComponents} = this.state

    return (
      <Grid padded divided columns='equal'>
        <Grid.Column>
          <Header as='h1' content='Form' />

          {ready &&
          <Form>
            {Object.keys(formComponents).map(value => {
              return <DCFormField key={value} properties={formComponents[value]} />
            })}
          </Form>
          }
        </Grid.Column>

        <Grid.Column>
          <Header as='h1' content='Test stuff' />

          <Header as='h3' content='All components' />

          <Checkbox label='Add warnings' checked={warning} onChange={this.handleCheckbox.bind(this, 'warning')} />

          <Divider hidden />

          <Checkbox label='Add errors' checked={error} onChange={this.handleCheckbox.bind(this, 'error')} />

          <Header as='h3' content='Dropdowns' />

          <Checkbox label='Wrong url' checked={urlError} onChange={this.handleDropdownErrors.bind(this, 'urlError')} />

          <Divider hidden />

          <Checkbox label='Network error' checked={networkError}
                    onChange={this.handleDropdownErrors.bind(this, 'networkError')} />
        </Grid.Column>
      </Grid>
    )
  }
}

export default App
