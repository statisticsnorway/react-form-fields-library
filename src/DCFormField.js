import React, { Component } from 'react'
import DCText from './components/DCText'
import DCBoolean from './components/DCBoolean'
import DCNumber from './components/DCNumber'
import DCRadio from './components/DCRadio'
import DCDate from './components/DCDate'
import DCDropdown from './components/DCDropdown'

class DCFormField extends Component {
  formComponents = {
    DCText: DCText,
    DCBoolean: DCBoolean,
    DCNumber: DCNumber,
    DCRadio: DCRadio,
    DCDate: DCDate,
    DCDropdown: DCDropdown
  }

  render () {
    const FormComponent = this.formComponents[this.props.properties.component]

    return <FormComponent {...this.props.properties} />
  }
}

export default DCFormField
