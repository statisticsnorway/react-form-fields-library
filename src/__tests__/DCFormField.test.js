import React from 'react'
import { shallow } from 'enzyme'

import DCFormField from '../DCFormField'
import { testFormComponents } from '../App'

describe('DCFormField', () => {
  it('Renders all the input components without error and only once', () => {
    Object.keys(testFormComponents).forEach(key => {
      const component = shallow(<DCFormField key={key} properties={testFormComponents[key]} />)

      expect(component.length).toEqual(1)
      expect(component).toMatchSnapshot()
    })
  })
})
