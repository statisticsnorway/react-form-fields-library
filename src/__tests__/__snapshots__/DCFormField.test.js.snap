// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DCFormField Renders all the input components without error 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DCFormField
    properties={
      Object {
        "component": "DCText",
        "description": "A description for this input",
        "displayName": "DCText",
        "error": "",
        "name": "myTextInput",
        "parent": "myForm",
        "required": true,
        "value": "",
        "warning": "",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "component": "DCText",
      "description": "A description for this input",
      "displayName": "DCText",
      "error": "",
      "name": "myTextInput",
      "parent": "myForm",
      "required": true,
      "value": "",
      "warning": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "component": "DCText",
        "description": "A description for this input",
        "displayName": "DCText",
        "error": "",
        "name": "myTextInput",
        "parent": "myForm",
        "required": true,
        "value": "",
        "warning": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`DCFormField Renders all the input components without error 2`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DCFormField
    properties={
      Object {
        "component": "DCBoolean",
        "description": "A description for this input",
        "displayName": "DCBoolean",
        "name": "myBooleanInput",
        "parent": "myForm",
        "value": false,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "component": "DCBoolean",
      "description": "A description for this input",
      "displayName": "DCBoolean",
      "name": "myBooleanInput",
      "parent": "myForm",
      "value": false,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "component": "DCBoolean",
        "description": "A description for this input",
        "displayName": "DCBoolean",
        "name": "myBooleanInput",
        "parent": "myForm",
        "value": false,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`DCFormField Renders all the input components without error 3`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DCFormField
    properties={
      Object {
        "component": "DCNumber",
        "description": "A description for this input",
        "displayName": "DCNumber",
        "error": "",
        "name": "myNumberInput",
        "parent": "myForm",
        "required": true,
        "value": "",
        "warning": "",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "component": "DCNumber",
      "description": "A description for this input",
      "displayName": "DCNumber",
      "error": "",
      "name": "myNumberInput",
      "parent": "myForm",
      "required": true,
      "value": "",
      "warning": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "component": "DCNumber",
        "description": "A description for this input",
        "displayName": "DCNumber",
        "error": "",
        "name": "myNumberInput",
        "parent": "myForm",
        "required": true,
        "value": "",
        "warning": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`DCFormField Renders all the input components without error 4`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DCFormField
    properties={
      Object {
        "component": "DCRadio",
        "description": "A description for this input",
        "displayName": "DCRadio",
        "error": "",
        "name": "myRadioInput",
        "options": Array [
          Object {
            "key": "thisOption",
            "text": "This option",
            "value": "thisOption",
          },
          Object {
            "key": "thatOption",
            "text": "That option",
            "value": "thatOption",
          },
          Object {
            "key": "somethingElse",
            "text": "Something else",
            "value": "somethingElse",
          },
        ],
        "parent": "myForm",
        "required": true,
        "value": "",
        "warning": "",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "component": "DCRadio",
      "description": "A description for this input",
      "displayName": "DCRadio",
      "error": "",
      "name": "myRadioInput",
      "options": Array [
        Object {
          "key": "thisOption",
          "text": "This option",
          "value": "thisOption",
        },
        Object {
          "key": "thatOption",
          "text": "That option",
          "value": "thatOption",
        },
        Object {
          "key": "somethingElse",
          "text": "Something else",
          "value": "somethingElse",
        },
      ],
      "parent": "myForm",
      "required": true,
      "value": "",
      "warning": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "component": "DCRadio",
        "description": "A description for this input",
        "displayName": "DCRadio",
        "error": "",
        "name": "myRadioInput",
        "options": Array [
          Object {
            "key": "thisOption",
            "text": "This option",
            "value": "thisOption",
          },
          Object {
            "key": "thatOption",
            "text": "That option",
            "value": "thatOption",
          },
          Object {
            "key": "somethingElse",
            "text": "Something else",
            "value": "somethingElse",
          },
        ],
        "parent": "myForm",
        "required": true,
        "value": "",
        "warning": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`DCFormField Renders all the input components without error 5`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DCFormField
    properties={
      Object {
        "component": "DCDate",
        "description": "A description for this input",
        "displayName": "DCDate",
        "error": "",
        "name": "myDateInput",
        "parent": "myForm",
        "required": true,
        "value": "",
        "warning": "",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "component": "DCDate",
      "description": "A description for this input",
      "displayName": "DCDate",
      "error": "",
      "name": "myDateInput",
      "parent": "myForm",
      "required": true,
      "value": "",
      "warning": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "component": "DCDate",
        "description": "A description for this input",
        "displayName": "DCDate",
        "error": "",
        "name": "myDateInput",
        "parent": "myForm",
        "required": true,
        "value": "",
        "warning": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`DCFormField Renders all the input components without error 6`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DCFormField
    properties={
      Object {
        "component": "DCDropdown",
        "description": "A description for this input",
        "displayName": "DCDropdown",
        "endpoints": Array [
          "https://metadata.ssbmod.net/data/Role/",
          "https://metadata.ssbmod.net/data/Agent/",
        ],
        "error": "",
        "multiSelect": false,
        "name": "myDropdownSingleSelectInput",
        "parent": "myForm",
        "required": true,
        "value": "",
        "warning": "",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "component": "DCDropdown",
      "description": "A description for this input",
      "displayName": "DCDropdown",
      "endpoints": Array [
        "https://metadata.ssbmod.net/data/Role/",
        "https://metadata.ssbmod.net/data/Agent/",
      ],
      "error": "",
      "multiSelect": false,
      "name": "myDropdownSingleSelectInput",
      "parent": "myForm",
      "required": true,
      "value": "",
      "warning": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "component": "DCDropdown",
        "description": "A description for this input",
        "displayName": "DCDropdown",
        "endpoints": Array [
          "https://metadata.ssbmod.net/data/Role/",
          "https://metadata.ssbmod.net/data/Agent/",
        ],
        "error": "",
        "multiSelect": false,
        "name": "myDropdownSingleSelectInput",
        "parent": "myForm",
        "required": true,
        "value": "",
        "warning": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`DCFormField Renders all the input components without error 7`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DCFormField
    properties={
      Object {
        "component": "DCDropdown",
        "description": "A description for this input",
        "displayName": "DCDropdown (multiSelect)",
        "endpoints": Array [
          "https://metadata.ssbmod.net/data/Agent/",
          "https://metadata.ssbmod.net/data/Role/",
        ],
        "error": "",
        "multiSelect": true,
        "name": "myDropdownMultipleSelectInput",
        "parent": "myForm",
        "required": true,
        "value": Array [],
        "warning": "",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "component": "DCDropdown",
      "description": "A description for this input",
      "displayName": "DCDropdown (multiSelect)",
      "endpoints": Array [
        "https://metadata.ssbmod.net/data/Agent/",
        "https://metadata.ssbmod.net/data/Role/",
      ],
      "error": "",
      "multiSelect": true,
      "name": "myDropdownMultipleSelectInput",
      "parent": "myForm",
      "required": true,
      "value": Array [],
      "warning": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "component": "DCDropdown",
        "description": "A description for this input",
        "displayName": "DCDropdown (multiSelect)",
        "endpoints": Array [
          "https://metadata.ssbmod.net/data/Agent/",
          "https://metadata.ssbmod.net/data/Role/",
        ],
        "error": "",
        "multiSelect": true,
        "name": "myDropdownMultipleSelectInput",
        "parent": "myForm",
        "required": true,
        "value": Array [],
        "warning": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
